### PR TITLE
xwayland: add drag'n'drop support

### LINKED ISSUE #295: xwayland clipboard sync and dnd
Whenever the wayland clipboard selection changes for a seat (selection-change event), we need to get the data source, read out the data from the fd and set the xwayland clipboard selection to its contents.

Whenever the X11 CLIPBOARD selection changes, we need to create a wayland data source, override the source methods, and set it on the seat.

A similar process will need to be done for the dnd selection (see the Weston xwayland implementation for guidance).

### PR DESCRIPTION
**Pull request by emersion**
- Add outgoing drag'n'drop support
- Fixes incr transfers (for big files)
- Refactor selection code

- [ ] Wayland → Xwayland
  - [x] Create DND window
  - [x] Send DND_ENTER
  - [x] Send DND_POSITION
  - [x] Send DND_DROP
  - [x] Receive DND_STATUS
  - [x] Send DND_LEAVE
  - [x] Receive DND_FINISHED
  - [x] Redesign `wlr_data_source`
  - [ ] Fix Firefox refusing text
  - [x] Fix Chromium crashing in non-incr transfer: support multiple transfers at the same time
  - [x] Only allow one transfer at a time
  - [x] Cleanup `wlr_xwm_selection_transfer` when transfer is finished
  - [x] Fix incr transfers
    - [x] Big picture from epiphany to gimp
    - [x] Big picture from chromium to lowriter
  - [x] Make sure I didn't broke clipboard stuff

Will be done in another PR:
- [ ] Xwayland → Wayland
  - [x] Figure out how the client knows about the Wayland window position: just draw a window that spans the whole screen
  - [ ] Figure out how to map X11 protocol to Wayland protocol
  - [ ] Receive DND_ENTER
  - [ ] Receive DND_POSITION
  - [ ] Receive DND_DROP
  - [ ] Send DND_STATUS
  - [ ] Receive DND_LEAVE
  - [ ] Send DND_FINISHED

Test clients:
* Firefox
* GTK+ (nice visual feedback)
* Chromium
* Gimp (has visual feedback too)

Fixes #295

### CODE REVIEW

**Comment by SirCmpwn** [comment]
Build failed

**Comment by SirCmpwn** [comment]
FYI

**Comment by SirCmpwn** [comment]
In case you were unaware

File: xwayland/xwm.c [hunk 1]
```diff
@@ -493,6 +507,26 @@ static void read_surface_net_wm_state(struct wlr_xwm *xwm,
...
+	xcb_get_atom_name_cookie_t name_cookie =
+		xcb_get_atom_name(xwm->xcb_conn, atom);
+	xcb_get_atom_name_reply_t *name_reply =
+		xcb_get_atom_name_reply(xwm->xcb_conn, name_cookie, NULL);
+	if (name_reply == NULL) {
+		return NULL;
+	}
+	size_t len = xcb_get_atom_name_name_length(name_reply);
+	char *buf = xcb_get_atom_name_name(name_reply); // not a C string
+	char *name = malloc((len + 1) * sizeof(char));
```
Comment (ascent12) [inline xwayland/xwm.c]: strndup

**Comment by emersion** [comment]
Is it possible to get this reviewed? I'll open an issue about the remaining Firefox issue.

**Comment by SirCmpwn** [comment]
LGTM but I can't test it until mods are fixed on the X11 backend

**Comment by emersion** [comment]
>LGTM but I can't test it until mods are fixed on the X11 backend

Merged the fix

**Comment by SirCmpwn** [comment]
Dragging image from epiphany to gimp (worked the first couple of times)

```
(gdb) bt
#0  0x00007f6d3abf2123 in wl_list_insert () at /usr/lib/libwayland-server.so.0
#1  0x00007f6d3ae55b8d in wlr_signal_emit_safe (signal=0x55cbcd46b7e0, data=0x55cbcd46b7a0)
    at ../util/signal.c:27
#2  0x00007f6d3ae382af in wlr_data_source_finish (source=0x55cbcd46b7a0)
    at ../types/wlr_data_device.c:1057
#3  0x00007f6d3ae38202 in data_source_resource_handle_destroy (resource=0x55cbcd46b830)
    at ../types/wlr_data_device.c:1038
#4  0x00007f6d3abedc12 in  () at /usr/lib/libwayland-server.so.0
#5  0x00007f6d3abf2619 in  () at /usr/lib/libwayland-server.so.0
#6  0x00007f6d3abedcff in wl_client_destroy () at /usr/lib/libwayland-server.so.0
#7  0x00007f6d3abeddb9 in  () at /usr/lib/libwayland-server.so.0
#8  0x00007f6d3abef9b2 in wl_event_loop_dispatch () at /usr/lib/libwayland-server.so.0
#9  0x00007f6d3abee14c in wl_display_run () at /usr/lib/libwayland-server.so.0
#10 0x000055cbcbfa53d1 in main (argc=1, argv=0x7ffd174b4c18) at ../rootston/main.c:81
```

**Comment by SirCmpwn** [comment]
Cannot reproduce that segfault after merging in master. I think this is good to merge.

**Pull request merged**